### PR TITLE
Refactor of validation

### PR DESCRIPTION
**Pull request by rynowak**

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/Validation/ValidationStateDictionary.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.AspNet.Mvc.ModelBinding.Validation
+{
+    public class ValidationStateDictionary : Dictionary<object, ValidationState>
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/Validation/ValidationStateDictionary.cs]: Implement IDictionary rather than subclassing

File: src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/Validation/ValidationStateDictionary.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            }
+
+            public int GetHashCode(object obj)
+            {
+                return RuntimeHelpers.GetHashCode(obj);
+            }
+        }
+    }
+
+    public class ValidationState
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/Validation/ValidationStateDictionary.cs]: Rename to ValidationStateEntry

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+
+        // Sorts validators based on whether or not they are 'required'. We want to run
+        // 'required' validators first so that we get the best possible error message.
+        private class ValidatorOrderComparer : IComparer<IModelValidator>
+        {
+            public static readonly ValidatorOrderComparer Instance = new ValidatorOrderComparer();
+
+            public int Compare(IModelValidator x, IModelValidator y)
+            {
+                var xScore = x.IsRequired ? 0 : 1;
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs]: Could we add a `Order` property on validators instead? It'd be in line with the way we do ordering in other parts of the application.

File: src/Microsoft.AspNet.Mvc.Core/Controllers/DefaultControllerActionArgumentBinder.cs [hunk 4]
```diff
@@ -82,20 +82,22 @@ public class DefaultControllerActionArgumentBinder : IControllerActionArgumentBi
...
-                    operationContext.ValidatorProvider,
-                    modelState,
-                    modelExplorer);
-
-                _validator.Validate(validationContext, modelBindingResult.ValidationNode);
+                if (modelBindingResult.Model == null && 
+                    modelState.GetValidationState(modelBindingResult.Key) != ModelValidationState.Valid)
+                {
+                    modelState.MarkFieldValid(modelBindingResult.Key);
+                }
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Controllers/DefaultControllerActionArgumentBinder.cs]: Take out this `if`.

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/CancellationTokenModelBinder.cs [hunk 5]
```diff
@@ -17,13 +17,7 @@ public Task<ModelBindingResult> BindModelAsync(ModelBindingContext bindingContex
...
             {
                 var model = bindingContext.OperationBindingContext.HttpContext.RequestAborted;
-                var validationNode =
-                    new ModelValidationNode(bindingContext.ModelName, bindingContext.ModelMetadata, model)
-                    {
-                        SuppressValidation = true,
-                    };
-
-                return ModelBindingResult.SuccessAsync(bindingContext.ModelName, model, validationNode);
+                return ModelBindingResult.SuccessAsync(bindingContext.ModelName, model);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/CancellationTokenModelBinder.cs]: Double check this is covered by an exclude filter

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/DictionaryModelBinder.cs [hunk 6]
```diff
@@ -79,12 +80,14 @@ public class DictionaryModelBinder<TKey, TValue> : CollectionModelBinder<KeyValu
...
-                if (valueResult.IsModelSet)
-                {
-                    validationNode.ChildNodes.Add(valueResult.ValidationNode);
-                }
+                keyMappings.Add(kvp.Key, convertedKey);
             }
 
+            bindingContext.ValidationState.Add(model, new ValidationState()
+            {
+                Strategy = new WierdExplicitIndexDictionaryValidationStrategy<TKey, TValue>(keyMappings, valueMetadata),
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/DictionaryModelBinder.cs]: `ShortFormDictionaryValidationStrategy`

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs]: I'd rather not even do that. The validator providers get a list, and they should be responsible for putting them in the right order.

Not part of this PR though

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/CancellationTokenModelBinder.cs (hunk 5 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/CancellationTokenModelBinder.cs]: This is already covered.

**Comment by ilmax** [comment]
@rynowak are there any chanches to allow async model validation? ;)

**Comment by rynowak** [comment]
@ilmax - not part of this change. Feel free to open an issue but it's something we're very unlikely to do because it increases the overhead significantly in a very hot path.

A more workable solution would to be implement whatever you need in an action filter, or replace the `IControllerActionArgumentBinder` which are both async.

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs [hunk 7]
```diff
@@ -576,30 +558,6 @@ private static void CallPropertyAddRange<TElement>(object target, object source)
             }
         }
 
-        // Returns true if validator execution adds a model error.
-        private static bool RunValidator(
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: This was unused.

**Comment by ilmax** [comment]
Thak you @rynowak for your quick response, I've already opened an issue in the past around a sync model validation. Performance is a fair argument to exclude the feature, but sometimes is useful (or even required) to hit some external resources to perform model validation. 
What I would like to have (as a opt-in feature maybe via a different package) is the ability to hook into the validation phase and write async logic without rewrite the whole validation components, is that possible in some way?

**Comment by rynowak** [comment]
> What I would like to have (as a opt-in feature maybe via a different package) is the ability to hook into the validation phase and write async logic without rewrite the whole validation components, is that possible in some way?

You can do this today by replacing `IControllerActionArgumentBinder` with your own implementation. You'd need to replace this code with your own validation system: https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNet.Mvc.Core/Controllers/DefaultControllerActionArgumentBinder.cs#L98

We'd be open to changes to make this kind of extensibility easier, but it's unlikely we'd put it in the box.

File: test/Microsoft.AspNet.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs [hunk 8]
```diff
@@ -107,13 +107,13 @@ private class Person6
             // Read-only collection should not be updated.
             Assert.Empty(boundModel.Address);
 
-            // ModelState (data is valid but is not copied into Address).
-            Assert.True(modelState.IsValid);
+            // ModelState (data is can't be validated).
+            Assert.False(modelState.IsValid);
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs]: @dougbu Let's talk about this case when you get to it.

**Comment by rynowak** [comment]
@dougbu  - updated with docs and such. My only outstanding TODO that I know of here is to write unit tests for each of the validation strategies.

File: test/Microsoft.AspNet.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs (hunk 8 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs]: Discussed in person, agreed that `Unvalidated` is fine.

File: src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/Validation/ValidationEntry.cs [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+            Metadata = metadata;
+            Key = key;
+            Model = model;
+        }
+
+        /// <summary>
+        /// The model prefix associated with <see cref="Model"/>.
+        /// </summary>
+        public string Key { get; private set; }
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/Validation/ValidationEntry.cs]: `{ get; }` For this and other properties.

File: src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/Validation/ValidationEntry.cs [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    /// Contains data needed for validating a child entry of a model object. See <see cref="IValidationStrategy"/>.
+    /// </summary>
+    public struct ValidationEntry
+    {
+        /// <summary>
+        /// Creates a new <see cref="ValidationEntry"/>.
+        /// </summary>
+        /// <param name="metadata">The <see cref="ModelMetadata"/> associated with <paramref name="model"/>.</param>
+        /// <param name="key">The model prefix associated with <paramref name="model"/>.</param>
+        /// <param name="model">the model object.</param>
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/Validation/ValidationEntry.cs]: `The model object.`

File: src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/Validation/ValidationEntry.cs [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+        public ValidationEntry( ModelMetadata metadata, string key, object model)
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/Validation/ValidationEntry.cs]: `(ModelMetadata ...`

File: src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/Validation/ValidationStateDictionary.cs [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+    /// <summary>
+    /// Used for tracking validation state to customize validation behavior for a model object.
+    /// </summary>
+    public class ValidationStateDictionary : 
+        IDictionary<object, ValidationStateEntry>, 
+        IReadOnlyDictionary<object, ValidationStateEntry>
+    {
+        private readonly Dictionary<object, ValidationStateEntry> _inner;
+
+        public ValidationStateDictionary()
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/Validation/ValidationStateDictionary.cs]: docs

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/CollectionModelBinder.cs [hunk 13]
```diff
@@ -252,17 +239,26 @@ private Task<CollectionResult> BindComplexCollection(ModelBindingContext binding
 
             return new CollectionResult
             {
-                ValidationNode = validationNode,
-                Model = boundCollection
+                Model = boundCollection,
+
+                // If we're working with a fixed set of indexes then this is the format like:
+                //
+                //  ?parameter.index=zero,one,two&parameter[zero]=0&&parameter[one]=1&parameter[two]=1...
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/CollectionModelBinder.cs]: `[two]=2...`

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/DefaultCollectionValidationStrategy.cs [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+    /// 
+    ///     In this case the elements of the collection are identified in the input data set by an incrementing
+    ///     integer index.
+    /// </example>
+    /// 
+    /// or:
+    /// 
+    /// <example>
+    ///     Model: IDictionary&lt;string, int&gt; 
+    ///     Query String: ?students[0].Name=Joey&amp;students[0].Age=8
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/DefaultCollectionValidationStrategy.cs]: Interesting, don't have to do `[0].Value.Name`?

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/DefaultCollectionValidationStrategy.cs [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNet.Mvc.ModelBinding.Validation
+{
+    /// <summary>
+    /// The default implementation of <see cref="IValidationStrategy"/> for a collection.
+    /// </summary>
+    /// <remarks>
+    /// This implemenation handles cases like:
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/DefaultCollectionValidationStrategy.cs]: implementation

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/DefaultCollectionValidationStrategy.cs (hunk 14 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/DefaultCollectionValidationStrategy.cs]: Yep, that's right. Overaggressive copy pasting of the examples.

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ExplicitIndexCollectionValidationStrategy.cs [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+    ///     class will yield two 'Student' objects with corresponding keys 'students[Joey]' and 'students[Katherine]'.
+    /// </example>
+    /// 
+    /// Using this key format, the enumerator enumerates model objects of type matching
+    /// <see cref="ModelMetadata.ElementMetadata"/>. The keys captured during model binding are mapped to the elements
+    /// in the collection to compute the model prefix keys.
+    /// </remarks>
+    public class ExplicitIndexCollectionValidationStrategy : IValidationStrategy
+    {
+        public ExplicitIndexCollectionValidationStrategy(IEnumerable<string> elementKeys)
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ExplicitIndexCollectionValidationStrategy.cs]: docs

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ExplicitIndexCollectionValidationStrategy.cs [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+        {
+            ElementKeys = elementKeys;
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ExplicitIndexCollectionValidationStrategy.cs]: This able to be `null`?

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs [hunk 18]
```diff
@@ -0,0 +1,356 @@
...
+            public static StateManager Recurse(
+                ValidationVisitor visitor,
+                string key,
+                ModelMetadata metadata,
+                object model,
+                IValidationStrategy strategy)
+            {
+                var recursifier = new StateManager(visitor, model);
+
+                visitor._container = visitor._model;
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs]: container is the model?

**Comment by NTaylorMullen** [comment]
Only found small issues. You gots a :shipit: from me :smile:

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ExplicitIndexCollectionValidationStrategy.cs (hunk 17 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ExplicitIndexCollectionValidationStrategy.cs]: Nope, that wouldn't work at all!

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs (hunk 18 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs]: Container is what we're inside. If we're validating a property, then model is the object we evaluated the property from.

Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs]: :+1:

File: src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ModelBindingContext.cs [hunk 19]
```diff
@@ -174,5 +178,11 @@ public ModelBindingContext()
         /// is eligible for model binding.
         /// </summary>
         public Func<ModelBindingContext, string, bool> PropertyFilter { get; set; }
+
+        /// <summary>
+        /// Gets or sets the <see cref="ValidationStateDictionary"/>. Used for tracking validation state to
+        /// customize validation behavior for a model object.
+        /// </summary>
+        public ValidationStateDictionary ValidationState { get; set; }
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ModelBindingContext.cs]: Why does this property have a setter? Is this just for the odd special-case almost-a-copy-constructor code in `CompositeModelBinder`?

File: src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/Validation/ValidationStateDictionary.cs [hunk 20]
```diff
@@ -0,0 +1,179 @@
...
+            get
+            {
+                ValidationStateEntry entry;
+                TryGetValue(key, out entry);
+                return entry;
+            }
+
+            set
+            {
+                _inner[key] = value;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/Validation/ValidationStateDictionary.cs]: In what scenario is a `null` `ValidationStateEntry` interesting? Put another way, why no `null` check here?

File: src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/Validation/ValidationStateDictionary.cs [hunk 21]
```diff
@@ -0,0 +1,179 @@
...
+                return ((IDictionary<object, ValidationStateEntry>)_inner).IsReadOnly;
+            }
+        }
+
+        /// <inheritdoc />
+        public ICollection<object> Keys
+        {
+            get
+            {
+                return ((IDictionary<object, ValidationStateEntry>)_inner).Keys;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/Validation/ValidationStateDictionary.cs]: Curious: Why not `_inner.Keys`?

File: src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/Validation/ValidationStateDictionary.cs [hunk 22]
```diff
@@ -0,0 +1,179 @@
...
+            return _inner.TryGetValue(key, out value);
+        }
+
+        /// <inheritdoc />
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return ((IDictionary<object, ValidationStateEntry>)_inner).GetEnumerator();
+        }
+        
+        private class ReferenceEqualityComparer : IEqualityComparer<object>
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/Validation/ValidationStateDictionary.cs]: Two copies of this code now. Perhaps the version in Mvc.Core should be public and in Mvc.Abstractions?

File: src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/Validation/ValidationStateDictionary.cs [hunk 23]
```diff
@@ -0,0 +1,179 @@
...
+        {
+            public static readonly ReferenceEqualityComparer Instance = new ReferenceEqualityComparer();
+
+            public new bool Equals(object x, object y)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/Validation/ValidationStateDictionary.cs]: `private`

File: src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/Validation/ValidationStateDictionary.cs [hunk 24]
```diff
@@ -0,0 +1,179 @@
...
+            {
+                return Object.ReferenceEquals(x, y);
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/Validation/ValidationStateDictionary.cs]: lowercase `object`

File: src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/Validation/ValidationStateEntry.cs [hunk 25]
```diff
@@ -0,0 +1,33 @@
...
+    /// <summary>
+    /// An entry in a <see cref="ValidationStateDictionary"/>. Records state information to override the default
+    /// behavior of validation for an object.
+    /// </summary>
+    public class ValidationStateEntry
+    {
+        /// <summary>
+        /// Gets or sets the model prefix associated with the entry.
+        /// </summary>
+        public string Key { get; set; }
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/Validation/ValidationStateEntry.cs]: See the need to modify an existing entry in `CompositeModelBinder`. But, pity this `class` can't be immutable, like `ValidationEntry`.

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/DefaultCollectionValidationStrategy.cs [hunk 26]
```diff
@@ -0,0 +1,123 @@
...
+namespace Microsoft.AspNet.Mvc.ModelBinding.Validation
+{
+    /// <summary>
+    /// The default implementation of <see cref="IValidationStrategy"/> for a collection.
+    /// </summary>
+    /// <remarks>
+    /// This implementation handles cases like:
+    /// <example>
+    ///     Model: IList&lt;Student&gt; 
+    ///     Query String: ?students[0].Age=8&students[1].Age=9
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/DefaultCollectionValidationStrategy.cs]: Invalid `&` here. You caught some of these just below and later in the PR but should be `&amp;` in all new examples.

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/DefaultObjectValidator.cs [hunk 27]
```diff
@@ -27,324 +21,49 @@ public class DefaultObjectValidator : IObjectModelValidator
...
-        {
-            public static readonly ValidatorOrderComparer Instance = new ValidatorOrderComparer();
-
-            public int Compare(IModelValidator x, IModelValidator y)
-            {
-                var xScore = x.IsRequired ? 0 : 1;
-                var yScore = y.IsRequired ? 0 : 1;
-                return xScore.CompareTo(yScore);
-            }
+            var metadata = model == null ? null : _modelMetadataProvider.GetMetadataForType(model.GetType());
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/DefaultObjectValidator.cs]: Shouldn't this get the `ModelMetadata` from the `ValidationStateDictionary`, when that's available?

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ExplicitIndexCollectionValidationStrategy.cs [hunk 28]
```diff
@@ -0,0 +1,128 @@
...
+{
+    /// <summary>
+    /// An implementation of <see cref="IValidationStrategy"/> for a collection bound using 'explict indexing'
+    /// style keys.
+    /// </summary>
+    /// <remarks>
+    /// This implemenation handles cases like:
+    /// <example>
+    ///     Model: IList&lt;Student&gt; 
+    ///     Query String: ?students.index=Joey,Katherine&amp;students[Joey].Age=8&students[Katherine].Age=9
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ExplicitIndexCollectionValidationStrategy.cs]: Another invalid `&`

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ExplicitIndexCollectionValidationStrategy.cs [hunk 29]
```diff
@@ -0,0 +1,128 @@
...
+                throw new ArgumentNullException(nameof(elementKeys));
+            }
+
+            ElementKeys = elementKeys;
+        }
+
+        /// <summary>
+        /// Gets the keys of collection elements that were used during model binding.
+        /// </summary>
+        public IEnumerable<string> ElementKeys { get; }
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ExplicitIndexCollectionValidationStrategy.cs]: `internal` with "for testing" comment?

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ShortFormDictionaryValidationStrategy.cs [hunk 30]
```diff
@@ -0,0 +1,135 @@
...
+            ModelMetadata valueMetadata)
+        {
+            KeyMappings = keyMappings;
+            _valueMetadata = valueMetadata;
+        }
+
+        /// <summary>
+        /// Gets the mapping from model prefix key to dictionary key.
+        /// </summary>
+        public IEnumerable<KeyValuePair<string, TKey>> KeyMappings { get; }
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ShortFormDictionaryValidationStrategy.cs]: `internal` with "for testing" comment?
